### PR TITLE
gazebo_plugins: moveit planning scene publisher

### PR DESCRIPTION
**Pull request by jbohren**
This gazebo plugin publishes all of the objects in a gazebo world to a moveit planning scene topic. It can either publish at a given rate, or be polled manually.

Example usage:

``` xml
      <gazebo>
        <plugin filename="libgazebo_ros_moveit_planning_scene.so" name="gazebo_ros_moveit_planning_scene">
          <topicName>/planning_scene</topicName>
          <sceneName>laboratory</sceneName>
          <robotName>my_robot</robotName>
          <updatePeriod>0.5</updatePeriod>
        </plugin>
      </gazebo>

```
- [ ] subscribe to published planning scene in order to make sure that models have already been added

Shoutout @davetcoleman @isucan

### CODE REVIEW

**Comment by jonbinney** [comment]
This sounds very cool....

**Comment by jbohren** [comment]
> This sounds very cool....

It's cool setting the update period, setting something spinning in gazebo, and then running MoveIt with re-planning.

**Comment by sachinchitta** [comment]
This is very cool - we have been talking about having something like this for quite a while. Thanks Jon!

File: gazebo_plugins/CMakeLists.txt [hunk 1]
```diff
@@ -137,7 +139,9 @@ add_dependencies(camera_synchronizer ${PROJECT_NAME}_gencfg)
 target_link_libraries(camera_synchronizer vision_reconfigure ${catkin_LIBRARIES})
 
 add_executable(pub_joint_trajectory_test test/pub_joint_trajectory_test.cpp)
-target_link_libraries(pub_joint_trajectory_test ${GAZEBO_LIBRARIES} ${catkin_LIBRARIES} ${Boost_LIBRARIES})
+# NOTE: QtGui and QtCore are required when using the GNU GoLD linker with Gazebo 2.2.2 which has poorly-specified so deps
+# More info here: https://bitbucket.org/osrf/gazebo/pull-request/1029/fix-second-order-dynamic-library/diff
+target_link_libraries(pub_joint_trajectory_test ${GAZEBO_LIBRARIES} ${catkin_LIBRARIES} ${Boost_LIBRARIES} QtGui QtCore)
```
Comment (davetcoleman) [inline gazebo_plugins/CMakeLists.txt:144]: Does this change belong in this PR? Seems to be unrelated. Did you mean to leave it?

File: gazebo_plugins/include/gazebo_plugins/gazebo_ros_moveit_planning_scene.h [hunk 2]
```diff
@@ -0,0 +1,163 @@
+/*
+ *  Gazebo - Outdoor Multi-Robot Simulator
+ *  Copyright (C) 2003  
+ *     Nate Koenig & Andrew Howard
```
Comment (davetcoleman) [inline gazebo_plugins/include/gazebo_plugins/gazebo_ros_moveit_planning_scene.h:4]: Seems to be the wrong license authors?

File: gazebo_plugins/include/gazebo_plugins/gazebo_ros_moveit_planning_scene.h [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+/// @{
+/** \defgroup GazeboRosMoveItPlanningScene Plugin XML Reference and Example
+
+  \brief Ros MoveIt Planning Scene Plugin.
+  
+  This is a model plugin which broadcasts MoveIt PlanningScene messages so
+  that the planning scene stays up-to-date with the world simulation.  This is
+  useful if you want to "fake" perfect perception of the environment.
+
+  Example Usage:
```
Comment (davetcoleman) [inline gazebo_plugins/include/gazebo_plugins/gazebo_ros_moveit_planning_scene.h:60]: Awesome

File: gazebo_plugins/include/gazebo_plugins/gazebo_ros_moveit_planning_scene.h [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+
+  /// \brief A pointer to the ROS node.  A node will be instantiated if it does not exist.
+  private: boost::scoped_ptr<ros::NodeHandle> rosnode_;
+  private: ros::Publisher planning_scene_pub_;
+           ros::ServiceServer publish_planning_scene_service_;
+
+  /// \brief A mutex to lock access to fields that are used in ROS message callbacks
+  private: boost::mutex mutex_;
+
+  /// \brief ROS Wrench topic name inputs
```
Comment (davetcoleman) [inline gazebo_plugins/include/gazebo_plugins/gazebo_ros_moveit_planning_scene.h:131]: Not a wrench

File: gazebo_plugins/include/gazebo_plugins/gazebo_ros_moveit_planning_scene.h [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+  private: std::string model_name_;
+
+  /// \brief for setting ROS name space
+  private: std::string robot_namespace_;
+
+  // Custom Callback Queue
+  private: ros::CallbackQueue queue_;
+  /// \brief Thead object for the running callback Thread.
+  private: boost::thread callback_queue_thread_;
+  /// \brief Container for the wrench force that this plugin exerts on the body.
```
Comment (davetcoleman) [inline gazebo_plugins/include/gazebo_plugins/gazebo_ros_moveit_planning_scene.h:145]: Bad comment.

**Comment by davetcoleman** [comment]
+1 I'm so glad you made this, thanks!

I tried reading through the exact details but didn't spend enough time, so I'll just ask instead. Is it removing and re-adding all the collision objects every update cycle, or only those that have changed? 

Looks good from my visual review!

**Comment by jbohren** [comment]
> I tried reading through the exact details but didn't spend enough time, so I'll just ask instead. Is it removing and re-adding all the collision objects every update cycle, or only those that have changed?

If an object appears/disappears, it calls ADD/REMOVE, and if it's been seen before, it just calls MOVE.

**Comment by jbohren** [comment]
Note that there's an issue if this gets loaded before the move_group node. I need to add some robustness there.

**Comment by cpaxton** [comment]
Add get_planning_scene service for MoveIt

File: gazebo_plugins/include/gazebo_plugins/gazebo_ros_moveit_planning_scene.h [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
```
Comment (hsu) [inline gazebo_plugins/include/gazebo_plugins/gazebo_ros_moveit_planning_scene.h:18]: also, please update to new license, see #225

**Comment by hsu** [comment]
@jbohren if you're OK with the license change, I can work on addressing some review comments?

File: gazebo_plugins/CMakeLists.txt (hunk 1 above)
Comment (hsu) [inline gazebo_plugins/CMakeLists.txt:144]: looking at https://bitbucket.org/osrf/gazebo/pull-request/1029/fix-second-order-dynamic-library/diff, is this still a necessary fix here?

Comment (jbohren) [inline gazebo_plugins/CMakeLists.txt:144]: Possibly not, I'd need to test it to make sure.

Comment (hsu) [inline gazebo_plugins/CMakeLists.txt:144]: Can I go ahead and merge this pull request without this modification, and propose adding back this change in a separate pull request?

Comment (hsu) [inline gazebo_plugins/CMakeLists.txt:144]: Ok, I think I've found the proper way to include QT dependencies.

basically,

```
find_package(Qt4 REQUIRED)
include ( ${QT_USE_FILE} )
add_definitions(${QT_DEFINITIONS})
```

and

```
target_link_libraries(pub_joint_trajectory_test ${GAZEBO_LIBRARIES} ${catkin_LIBRARIES} ${Boost_LIBRARIES} ${QT_LIBRARIES})
```

Will merge that, please test and let me know if we need to perform further fix.  Thanks.

**Comment by davetcoleman** [comment]
I don't understand why this PR was merged manually - how does that even happen?

Also, it seems my comments from my code review (all very minor changes) where not addressed.

**Comment by hsu** [comment]
@davetcoleman 
I merged it manually. (when the pull request was open, there's an option near the merge-button to merge manually).

Sorry, did I fail to address a comment from you?

**Comment by davetcoleman** [comment]
On the [Files Changed](https://github.com/ros-simulation/gazebo_ros_pkgs/pull/192/files) tab, there are several small commenting/naming bugs I found and commented that should be fixed (since I took the time to review it :-) )

**Comment by jbohren** [comment]
> On the Files Changed tab, there are several small commenting/naming bugs I found and commented that should be fixed (since I took the time to review it :-) )

Dave, I've got an undergraduate who started in our lab this week who I'm going to have go through and address the comments. I'll have her create another PR to address them.

**Comment by hsu** [comment]
i see, the changes do not show up in the files changed tab, but they were addressed during merge... at least i thought i did.  i'll double check the hydro-devel branch and see if i missed any?  thanks!

**Comment by davetcoleman** [comment]
> I've got an undergraduate who started in our lab this week

I'm working on getting one of those too ;-)

**Comment by hsu** [comment]
@davetcoleman 

I think the issues have  been addressed:
- license updated and wrench comments were removed from the [header](https://github.com/ros-simulation/gazebo_ros_pkgs/blob/hydro-devel/gazebo_plugins/include/gazebo_plugins/gazebo_ros_moveit_planning_scene.h)
- qt related issue was mostly addressed in [CMakeLists.txt](https://github.com/ros-simulation/gazebo_ros_pkgs/blob/hydro-devel/gazebo_plugins/CMakeLists.txt), pending more feedback.

was there anything else I missed? (besides the undergrad you wanted) :)

Thanks!

**Comment by davetcoleman** [comment]
I think that's it, perfect.

I'll keep working on the undergrad here in Colorado.

@jbohren you should email the moveit mailing list about this new plugin, its pretty cool

**Comment by davetcoleman** [comment]
> I'll keep working on the undergrad here in Colorado.

Totally off topic, but I believe I got one today :-p

**Comment by hsu** [comment]
:+1:
